File: Src/java/model/src/main/java/org/hl7/cql/model/BaseDataType.java
package org.hl7.cql.model;

public abstract class BaseDataType implements DataType {
    protected BaseDataType() {
        this(DataType.ANY);
    }

    protected BaseDataType(DataType baseType) {
        this.baseType = baseType != null ? baseType : DataType.ANY;
    }

    private final DataType baseType;

    @Override
    public DataType getBaseType() {
        return baseType;
    }

    @Override
    public String toLabel() {
        return toString();
    }

    @Override
    public boolean isSubTypeOf(DataType other) {
        DataType currentType = this;
        while (currentType != null) {
            if (currentType.equals(other)) {
                return true;
            }
            currentType = currentType.getBaseType();
        }

        return false;
    }

    @Override
    public boolean isSuperTypeOf(DataType other) {
        while (other != null) {
            if (equals(other)) {
                return true;
            }
            other = other.getBaseType();
        }

        return false;
    }

    /**
     * @param other
     * @return The first supertype of this type that is also a supertype of other
     */
    @Override
    public DataType getCommonSuperTypeOf(DataType other) {
        DataType currentType = this;
        while (currentType != ANY) {
            if (currentType.isSuperTypeOf(other)) {
                return currentType;
            }
            currentType = currentType.getBaseType();
        }

        return ANY;
    }

    // Note that this is not how implicit/explicit conversions are defined, the notion of
    // type compatibility is used to support implicit casting, such as casting a "null"
    // literal to any other type, or casting a class to an equivalent tuple.
    @Override
    public boolean isCompatibleWith(DataType other) {
        // A type is compatible with a choice type if it is a subtype of one of the choice types
        if (other instanceof ChoiceType) {
            for (DataType choice : ((ChoiceType) other).getTypes()) {
                if (this.isSubTypeOf(choice)) {
                    return true;
                }
            }
        }

        return this.equals(other); // Any data type is compatible with itself
    }
}

File: Src/java/model/src/main/java/org/hl7/cql/model/BaseDataType.kt
package org.hl7.cql.model

abstract class BaseDataType protected constructor(baseType: DataType? = DataType.ANY) : DataType {
    override val baseType: DataType = baseType ?: DataType.ANY

    override fun toLabel(): String {
        return toString()
    }

    override fun isSubTypeOf(other: DataType): Boolean {
        var currentType: DataType = this
        while (currentType != DataType.ANY) {
            if (currentType == other) {
                return true
            }
            currentType = currentType.baseType
        }

        return currentType == other
    }

    override fun isSuperTypeOf(other: DataType): Boolean {
        var currentType = other
        while (currentType != DataType.ANY) {
            if (this == currentType) {
                return true
            }
            currentType = currentType.baseType
        }

        return this == currentType
    }

    /**
     * @param other
     * @return The first supertype of this type that is also a supertype of other
     */
    override fun getCommonSuperTypeOf(other: DataType): DataType {
        var currentType: DataType = this
        while (currentType != DataType.ANY) {
            if (currentType.isSuperTypeOf(other)) {
                return currentType
            }
            currentType = currentType.baseType
        }

        return DataType.ANY
    }

    // Note that this is not how implicit/explicit conversions are defined, the notion of
    // type compatibility is used to support implicit casting, such as casting a "null"
    // literal to any other type, or casting a class to an equivalent tuple.
    override fun isCompatibleWith(other: DataType): Boolean {
        return when {
            // Any data type is compatible with itself
            this == other -> true
            // A type is compatible with a choice type if it is a subtype of one of the choice types
            other is ChoiceType -> other.types.any { this.isSubTypeOf(it) }
            else -> false
        }
    }
}
